src/type_methods.jl

```diff
@@ -0,0 +1,106 @@
+
+# build expression for tuples of the parameterizations data
+function expr_gen(typed, fields)
+    expr = :(())  # build an expression for a tuple
+    acc = typed ? typed_acc : untyped_acc
+    append!(expr.args, [acc(field) for field in fields])
+    return expr
+end
+untyped_acc(i) = isa(i, Int) ? :(X[$(i)]) : :(X.$(i))
+typed_acc(i) = isa(i, Int) ? :(T(X[$(i)])) : :(T(X.$(i)))
+
+
+#
+# function to create a code block to convert to and from Vec and Vector rtypes
+#
+function add_methods(rot_type)
+
+    #
+    # Build expressions for building typed and untyped tuples using an Integer Index
+    #
+    untyped_indexed_expr = expr_gen(false, 1:numel(rot_type))
+    typed_indexed_expr = expr_gen(true, 1:numel(rot_type))
+
+    #
+    # Build expressions for building typed and untyped tuples using a field
+    #
+
+    fields = fieldnames(rot_type)
+    untyped_field_expr = (rot_type <: Mat) ? untyped_indexed_expr : expr_gen(false, fields[1:numel(rot_type)])
+    typed_field_expr = (rot_type <: Mat) ? typed_indexed_expr : expr_gen(true, fields[1:numel(rot_type)])
+
+
+    # the code block to return
+    qb = quote end
+
+    #
+    # convert to / from a mutable vector
+    #
+    qn = quote
+
+        # convert to a mutable vector
+        @inline convert(::Type{Vector}, X::$(rot_type)) = vcat($(untyped_field_expr.args...))
+        @inline convert{T <: Real}(::Type{Vector{T}}, X::$(rot_type)) = vcat($(typed_field_expr.args...))
+        @inline call{T <: Vector}(::Type{T}, X::$(rot_type)) = convert(T, X)
+
+        # convert from a mutable vector
+        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::Vector{U}) = T($(untyped_indexed_expr.args...))
+        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::Vector{U}) = convert(T, X)
+
+        # convert to an immutable vector
+        @inline convert(::Type{Vec}, X::$(rot_type)) = Vec($(untyped_field_expr.args...))
+        @inline convert{T <: Real}(::Type{Vec{$(numel(rot_type)), T}}, X::$(rot_type)) = Vec($(typed_field_expr.args...))
+        @inline call(::Type{Vec}, X::$(rot_type)) = convert(Vec, X)
+        @inline call{T <: Real}(::Type{Vec{$(numel(rot_type)), T}}, X::$(rot_type)) = convert(Vec{$(numel(rot_type)), T}, X)
+
+        # convert from an immutable vector
+        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::FixedVector{$(numel(rot_type)), U}) = T($(untyped_indexed_expr.args...))
+        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::FixedVector{$(numel(rot_type)), U}) = convert(T, X)
+
+        # convert to a tuple
+        # @inline tuple(X::$(rot_type)) = tuple($(untyped_field_expr.args...))
+
+        # convert from a tuple
+        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::NTuple{$(numel(rot_type)), U}) = T($(untyped_indexed_expr.args...))
+        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::NTuple{$(numel(rot_type)), U}) = convert(T, X)
```

Comment:
It seems a somewhat questionable to be adding so many generic constructors and conversions to types which aren't defined in this module, eg `Quaternion` and `RotMatrix` at least.

Is the idea that you want to flatten the storage so you can stick it into a larger array for optimization?  It might be better to cut the methods back here, perhaps implement `vec` to flatten the types?  (`vec()` for FSAs should go into FixedSizeArrays, presumably returning an appropriate `Vec` from any other FSA).